### PR TITLE
Premove

### PR DESCRIPTION
**Pull request by LKL1235**
premove function which https://github.com/trevorbayless/cli-chess/issues/27 discussed

### CODE REVIEW

**Comment by trevorbayless** [comment]
Thanks for the PR! I haven't had time to look at it quite yet, but there are some [lint errors](https://github.com/trevorbayless/cli-chess/actions/runs/9168578213/job/25226917881?pr=29) that need to be addressed.

You can run `flake8` locally to verify you have fixed the errors. To do so, in your `venv` install cli-chess with `dev` dependencies by using the following command: `pip install -e .[dev]`. You should then navigate to `cli-chess/src/cli_chess` and run `flake8`.

File: src/cli_chess/modules/premove/premove_view.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from cli_chess.modules.premove import PremovePresenter
+
+
+class PremoveView:
+    def __init__(self, presenter: PremovePresenter, premove: str|None = None):
+        self.presenter = presenter
+        self.premove = premove
+        self.update(premove)
```
Comment (trevorbayless) [inline src/cli_chess/modules/premove/premove_view.py:28]: Unnecessary call to update in initialization

File: src/cli_chess/core/game/offline_game/offline_game_model.py [hunk 2]
```diff
@@ -49,12 +50,44 @@ def make_move(self, move: str):
...
+
+            except Exception:
+                raise
+        else:
+            log.warning("Attempted to make a move in a game that's not in progress")
+            raise Warning("Game has already ended")
+        
+    def make_premove(self, move: str):
+        """Verify move and Set board model premove"""
+        if self.game_in_progress:
```
Comment (trevorbayless) [inline src/cli_chess/core/game/offline_game/offline_game_model.py:68]:
```suggestion
        if self.game_in_progress and not self.is_my_turn():
```

File: src/cli_chess/core/game/offline_game/offline_game_model.py [hunk 3]
```diff
@@ -49,12 +50,44 @@ def make_move(self, move: str):
...
+            try:
+                if self.board_model.board.is_game_over():
+                    self.game_in_progress = False
+                    raise Warning("Game has already ended")
+                if self.board_model.get_premove() is not None:
+                    raise Warning("You already have a premove set")
+                if self.is_my_turn():
+                    raise Warning("It is your turn, cant not make a premove")
```
Comment (trevorbayless) [inline src/cli_chess/core/game/offline_game/offline_game_model.py:76]: Remove as this should not be hit.
```suggestion
                if self.is_my_turn():
                    raise Warning("It is your turn, cant not make a premove")
```

**Review by trevorbayless — CHANGES_REQUESTED** [review]
Thanks again for the PR! I've added some comments to be addressed.

A couple additions to the review:
- The "PreMove" text is not appearing on online games.
- Premoves currently will kill an entire online game session if the pre-move is accepted but is then invalid after the opponent moves. Example: Starting as white, play e4 and premove e5. If the opponent plays e5 the game is then in a bad state. You should be able to replicate this easily in an online game against Stockfish. When quitting cli-chess after this happens, you will see exceptions in the console.

**Comment by LKL1235** [comment]
> Thanks again for the PR! I've added some comments to be addressed.
> 
> A couple additions to the review:
> 
> * The "PreMove" text is not appearing on online games.
> * Premoves currently will kill an entire online game session if the pre-move is accepted but is then invalid after the opponent moves. Example: Starting as white, play e4 and premove e5. If the opponent plays e5 the game is then in a bad state. You should be able to replicate this easily in an online game against Stockfish. When quitting cli-chess after this happens, you will see exceptions in the console.

Thanks for guide, I did fix them and commit

**Review by trevorbayless — APPROVED** [review]

**Pull request merged**